File: BinaryTree/CPP/leaf_similar_trees.cpp
#include <vector>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode() : val(0), left(nullptr), right(nullptr) {}
     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        vector<int> l, r;
        helper(root1, l);
        helper(root2, r);
        if (l.size() != r.size()) return false;
        for (int i = 0; i < l.size(); i++) {
            if (l[i] != r[i]) return false;
        }
        return true;
    }

    void helper(TreeNode* root, vector<int>& ans) {
        if (root->left != NULL) helper(root->left, ans);
        if (root != NULL && root->left == NULL && root->right == NULL) ans.push_back(root->val);
        if (root->right != NULL) helper(root->right, ans);
    }
};

File: BinaryTree/Python/leaf_similar_trees.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        l, r = [], [] 
        def helper(root, ans):
            if root.left: helper(root.left, ans)
            if root and not root.left and not root.right: ans.append(root.val) 
            if root.right: helper(root.right, ans)
        helper(root1, l)
        helper(root2, r)
        return l == r